doc/en/man/duende.8
.\" Do *not* edit this file; it was automatically generated by ej2man
.\" Look for a name.ej file with the same name as this filename
.\"
.\" Process this file with the following (replace filename.1)
.\" preconv < filename.1 | nroff -man -Tutf8
.\"
.\" Last updated 2022-11-28
.\"
.TH DUENDE 8 "duende" "January 2003" "duende"
.\" We don't want hyphenation (it's too ugly)
.\" We also disable justification when using nroff
.\" Due to the way the -mandoc macro works, this needs to be placed
.\" after the .TH heading
.hy 0
.if n .na
.\"
.\" We need the following stuff so that we can have single quotes
.\" In both groff and other UNIX *roff processors
.if \n(.g .mso www.tmac
.ds aq \(aq
.if !\n(.g .if '\(aq'' .ds aq \'

  
.SH "NAME"
.PP
duende - run a child process as a daemon 
.SH "DESCRIPTION"
.PP

.B "duende"
is a tool for users of init systems which require programs to provide 
their own daemonization. 
.B "duende"
is not needed with systemd and other init systems which can provide 
daemonization for network services. 
.B "duende"
makes a given child process a daemon. The standard output and standard 
error of the child process is logged via syslog() with a priority of 
LOG_INFO. 
.SH "SYSTEMD"
.PP

.B "duende"
should not be used if one uses <rr>systemd as the init process. 
Instead, files like this should be added to /etc/systemd/system

.nf
# Place this file here: 
# /etc/systemd/system/maradns.service 
# Then 
# systemctl enable maradns 
# systemctl start maradns 
# To view log 
# journalctl -u maradns 
[Unit] 
 After=network-online.target 
 ConditionPathExists=/usr/local/bin/maradns 
 Description=MaraDNS 
 Wants=network-online.target 
 
[Service] 
 ExecStart=/usr/local/bin/maradns 
 StandardOutput=journal 
 TimeoutSec=0 
 Type=exec 
 
[Install] 
 WantedBy=multi-user.target 
.fi

When installed with make install, MaraDNS will look for systemd files, 
and install MaraDNS and Deadwood startup files. 
.SH "USAGE"
.PP

.B "duende"
(--pid=/path/to/file) child_process [ all subsequent arguments passed 
on to child ] 
.SH "DETAILS"
.PP
When 
.B "duende"
is invoked, it spawns two processes. In addition to spawning the 
daemonized child process, 
.B "duende"
also spawns a process which reads and logs the standard output of the 
daemonized process. The parent process stays alive so as to monitor the 
daemonized process. If the optional --pid argument is supplied, 
.B "duende"
will write its PID to the file specified by the argument. It is an 
error to supply the --pid argument without an equal sign and file name. 
.PP
.B "duende"
requires a blank directory named /etc/maradns/logger to run. 
.PP
Should the parent duende process a HUP signal, 
.B "duende"
will restart the child process. Should the daemonized or logging 
process received an untrapped HUP signal or exit with an exit code of 
8, 
.B "duende"
will restart the process. Should the daemonized or logging process exit 
for any other reason, 
.B "duende"
will send the logger process a TERM signal and exit. Should the duende 
parent process receive a TERM or INT signal, 
.B "duende"
sends all of its children TERM signals, then exits. 
.PP
The duende process must be started as the superuser; this is because 
Duende\(aqs intended child processes (maradns and zoneserver) need to 
bind to privileged ports, and because duende uses a setuid() call to 
change the user ID of the logging process to the user with ID 707. 
.SH "LOGGING"
.PP

.B "duende"
uses the syslog() facility to log the standard output of the program 
that it invokes. The name of the program (in other words, the "ident" 
given to openlog()) is the full path of the first argument given to 
.BR "duende" "."
All messages created by the child process are sent to syslog() with a 
priority of LOG_INFO and a "facility" of LOG_DAEMON (daemon.info in 
/etc/syslog.conf); since daemon.info messages are not logged by default 
in FreeBSD, on FreeBSD systems messages generated by the child process 
are logged with a priority of LOG_ALERT and a "facility" of LOG_DAEMON 
(daemon.alert in /etc/syslog.conf). Should duende itself encounter an 
error, it will send messages to syslog() with a priority of LOG_ALERT. 
.PP
For example, suppose one invokes duende thusly:

.nf
	duende /usr/local/sbin/maradns 
.fi

If invoked thusly, duende will log all messages with the "ident" 
(program name) of "/usr/local/sbin/maradns". If this is not desired, 
invoke duende with something like:

.nf
	export PATH=$PATH:/usr/local/sbin 
	duende maradns 
.fi

This will log messages with a (more sensible) "ident" of maradns. 
.PP
Note: If a non-POSIX Bourne shell (such as csh, es, rc, or fish) is 
used to invoke MaraDNS, the above syntax needs to be changed. 
.PP
Also, the directory /etc/maradns/logger, while used by duende, is not 
used to store any log messages. That is unless, for some reason, one 
configures syslog to store messages there. 
.SH "EXAMPLES"
.PP
Using duende to start maradns, where the mararc file is /etc/mararc.2

.nf
	duende maradns -f /etc/mararc.2 
.fi

Using duende to start zoneserver, where the mararc file is 
/etc/mararc.4

.nf
	duende zoneserver -f /etc/mararc.4 
.fi
.SH "BUGS"
.PP

.B "Duende"
assumes that all of its children are well-behaved, eating their 
vegetables, going to bed when told, and terminating when receiving a 
TERM signal. 
.SH "SEE ALSO"
.PP

.BR "maradns(8)" ","

.B "syslog(3)"

.br
http://www.maradns.org 
.SH "LEGAL DISCLAIMER"
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHORS \(aq\(aqAS IS\(aq\(aq AND ANY 
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR 
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
.SH "AUTHOR"
.PP
Duende and this man page are written by Sam Trenholme. D Richard Felker 
III provided some invaluable assistance with the piping code which 
.B "duende"
uses.  



CHANGELOG
MaraDNS changelog

   maradns-3.5.0020:
   This is a stable release of MaraDNS:

     * New parameter: source_ip4, to specify the source IP when sending a
       query to an upstream or authoritative DNS server.
     * Makefile.centos8 file for Deadwood renamed Makefile.Ubuntu2004 (i.e.
       it’s now a Makefile for Ubuntu 20.04)

     (2021-05-??)

   maradns-3.5.0019:
   This is a stable release of MaraDNS:

     * One line change to zoneserver.c to make it work better with systemd
     * Synthetic IP generator example (e.g. 10.1.2.3.ip4.internal 
       resolves to 10.1.2.3) added to coLunacyDNS documentation

     (2021-03-16)

   maradns-3.5.0018:
   This is a stable release of MaraDNS:

     * coLunacyDNS updated to 1.0.010 (Ubuntu 20.04 testing found a
       minor select() bug)
     * Since RedHat has broken their pinky promise to support CentOS 8
       until 2029, MaraDNS has moved from CentOS 8 to Ubuntu 20.04 LTS
       (RPM .spec files removed; Docker container now runs Ubuntu 20.04)

     (2020-12-20)

   maradns-3.5.0017:
   This is a stable release of MaraDNS:

     * coLunacyDNS update: We can now specify the “Authoritative”
       and “Recursion available” flags in the reply.
     * coLunacyDNS update: We can now specify a TTL for the reply,
       to be anywhere from 0 seconds (do not cache) to a little over
       90 days.
     * coLunacyDNS version updated to be 1.0.009.

     (2020-10-02)

   maradns-3.5.0016:
   This is a stable release of MaraDNS:

     * Unstable mmLunacyDNS code removed from tree (coLunacyDNS can do
       anything mmLunacyDNS could do)
     * coLunacyDNS bug fixes: We return with an error if the Lua code
       attempts to return an invalid IPv4 address to the client (before,
       the code incorrectly returned 255.255.255.255)
     * coLunacyDNS returns helpful errors if the processQuery return 
       value is invalid in various ways.
     * coLunacyDNS now has 100%* testing coverage. *Some sanity tests
       which protect coLunacyDNS from security threats which can not
       be readily reproduced are disabled in testing mode.
     * coLunacyDNS is now at version 1.0.008

     (2020-09-01)

   maradns-3.5.0015:
   This is a stable release of MaraDNS:

     * coLunacyDNS is now a stable release (1.0.007) with a full
       SQA testing suite and well over 90% code coverage in its tests.
     * mmLunacyDNS has been removed; coLunacyDNS can do everything
       mmLunacyDNS could do, and this saves me the bother of
       maintaining two code bases.
     * askmara now compiles with IPv6 support (the code has been
       there, but is finally getting enabled)

     (2020-08-29)

   maradns-3.5.0014:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS’s Lua script can now specify IPv6 addresses in
       standard “colon” format, e.g. co1Data="2001:db8::1"
     * coLunacyDNS now handles ANY (and HINFO) queries as per RFC8482
     * coLunacyDNS documentation updates: Various cleanup.  Also,
       the example coLunacyDNS .lua files now return “not there”
       when we ask for a hostname which does not have a given record.

     (2020-08-20)

   maradns-3.5.0013:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS now has support for binding to IPv6 addresses.  Both
       the *NIX (Linux) and the Windows32 binary can bind to an IPv6
       socket.
     * Some other bug fixes and cleanup, mainly with coLunacyDNS.

     (2020-08-19)

   maradns-3.5.0012:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable): 

     * mmLunacyDNS security fix: We now use a secure hash compression
       function (HalfSipHash-1-3) for string hashing.
     * coLunacyDNS: hash compression function updated from 64-bit
       SipHash-2-4 to 32-bit HalfSipHash-1-3.  Compile time warnings
       removed from code.
     * lunacy: The code by default now uses HalfSipHash-1-3 for string
       hash compression.  Default compile optimization is now -O3
 
     (2020-08-12)

   maradns-3.5.0011:
   This is a stable release of MaraDNS:

     * min_ttl parameter added; this is the minimum time we keep a 
       record in the cache (in seconds)
     * Deadwood now compiles with IPv6 support by default.  For
       systems without IPv6 support, -DNOIP6 can be set when
       compiling Deadwood.
     * Automated tests now all run inside of Podman (Docker) container
       and all pass.  Tests are now completely automated, and can run
       from cron (and can be adapted to run inside Jenkins). 

     (2020-08-10)

   maradns-3.5.0010:
   This is a stable release of MaraDNS:

     * Hotfix: coLunacyDNS no longer fails after 20 calls to 
       processQuery() (we now properly clean the main stack before 
       calling processQuery() in a co-routine).
     * Security update: MaraDNS, Deadwood, and Duende now default to
       the user ID 707 instead of 99/66.  This minimizes the chances
       of the user used by MaraDNS being used by other processes, 
       which could be a security leak under some circumstances.  The
       problem with running multiple services as "nobody" is that
       the "nobody" account is only as secure as the least secure
       service running as that account.
     * coLunacyDNS feature update: coLunacyDNS can now open and
       read files (for security reasons, only in the same directory
       coLunacyDNS is running in).  In addition, the code to implement 
       IPv6 sockets is well under way.

     (2020-08-06)

   maradns-3.5.0009:
   This is a stable release of MaraDNS:

     * Add new program: coLunacyDNS.  This is a DNS server which runs a
       Lua function every time it gets a DNS query.  It uses Lua
       threads ("co-routines") to have a function which can get a
       DNS packet from an upstream server and return the result for

deadwood-github/tools/duende.c
/* Copyright (c) 2003-2008 Sam Trenholme
 *
 * TERMS
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * This software is provided 'as is' with no guarantees of correctness or
 * fitness for purpose.
 */

/* I would like to thank D. Richard Felker III for his invaluable help
   with debugging how the pipes are set up here */

/* This program is a helper application which does the following:

   It starts the maradns process

   It syslog()s maradns' output via a chroot'd unprivledged child process

   If maradns exits with a code of 8, this means that she was given a
   HUP signal.  Restart the MaraDNS process

   If this process gets a HUP signal, restart the MaraDNS process

 */

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

/* The UID that the Duende logging process uses.  CHANGE THE DUENDE MAN
   PAGE IF YOU CHANGE THIS VALUE (same general process as changing the
   mararc man page; the source file for the duende man page is duende.ej) */
#define DUENDE_LOGGER_UID 707

/* The directory that Duende runs in.  This directory has to exist for
   Duende to be able to run.  Again, IF YOU CHANGE THIS, CHANGE THE
   DUENDE MAN PAGE */
#define DUENDE_CHROOT_DIR "/etc/maradns/logger"

int got_hup_signal = 0;
int got_term_signal = 0;

/* If we get a HUP signal set the flag so we can restart the MaraDNS child
   process */
void handle_hup() {
    got_hup_signal = 1;
    return;
    }

void handle_term() {
    got_term_signal = 1;
    return;
    }

/* Helper process which syslogs stuff from either MaraDNS' stdout or stderr. */

void log_helper(char *name,int stdout_fd) {
    char out_buf[1024];

    /* We can't use our signal handlers because fgets is blocking */
    signal(SIGTERM,SIG_DFL);
    signal(SIGHUP,SIG_DFL);

    /* Open up the sys log */
    openlog(name,0,LOG_DAEMON);

    /* Drop all privileges */
    if(chdir(DUENDE_CHROOT_DIR) != 0) {
       syslog(LOG_ALERT,"Can not enter chroot directory %s",DUENDE_CHROOT_DIR);
        syslog(LOG_ALERT,"%s","We can not log daemon output");
        printf("Fatal error logging; read syslog\n");
        printf("%s directory required to exist\n",DUENDE_CHROOT_DIR);
        exit(1);
        }
    if(setuid(DUENDE_LOGGER_UID) != 0) {
        syslog(LOG_ALERT,"%s%d","Can not change UID to ",DUENDE_LOGGER_UID);
        syslog(LOG_ALERT,"%s","We can not log daemon output");
        printf("Fatal error logging; read syslog\n");
        exit(1);
        }

    /* Log both stdout and stderr */
    dup2(stdout_fd,0);
    for(;;) {
        if(fgets(out_buf,1020,stdin) == out_buf)
#ifdef __FreeBSD__
            /* FreeBSD doesn't log daemon.info messages by default; while
             * this can be changed by editing /etc/syslog.conf, it *is*
             * an issue that can cause confusion */
            syslog(LOG_ALERT,"%s",out_buf);
#else /* __FreeBSD__ */
            syslog(LOG_INFO,"%s",out_buf);
#endif /* __FreeBSD__ */
        }

    syslog(LOG_ALERT,"log_helper process terminating");
    exit(0);

    }


/* If a child exits, see if the child exited with a code of 8 or received
   a HUP signal.  In either of these cases, restart the child daemon and the
   (if needed) logger process).  Otherwise, exit */

void handle_child_exited(int exit_status, pid_t alive, pid_t exited) {
        if(WIFEXITED(exit_status)) { /* Exit with exit code */
            if(WEXITSTATUS(exit_status) != 8) { /* Anything but HUP */
                kill(alive,SIGTERM);
                syslog(LOG_ALERT,"Child exited with status %d",exit_status);
                exit(WEXITSTATUS(exit_status));
                }
            }
        if(WIFSIGNALED(exit_status)) { /* Got signal */
            if(WTERMSIG(exit_status) != SIGHUP) {
                syslog(LOG_ALERT,"Child got signal %d",exit_status);
                kill(alive,SIGTERM);
                exit(1);
                }
            }
       /* If you somehow stop the child daemon, we go bye bye */
       if(WIFSTOPPED(exit_status)) {
            syslog(LOG_ALERT,"Child stopped");
            kill(exited,SIGTERM);
            kill(alive,SIGTERM);
            exit(2);
            }
        /* Clean up the system logging */
        syslog(LOG_ALERT,"Cleaning up system logging");
        kill(alive,SIGTERM);
    }

/* The main process forks off the child.  Right now, I will just have
   it fork off the MaraDNS process, hardwired as /usr/sbin/maradns,
   directing her standard output to
   /dev/null.  The revision of this file will correctly handle Mara's
   output
 */

int main(int argc, char **argv) {
    int exit_status;
    pid_t pid, log_pid;
    int stream1[2]; /* Used for piping */
    int exec_argv_offset = 1; /* Also used to determine PID writing */
    if(argv[0] == NULL || argv[1] == NULL) {
        printf("Usage: duende (--pid=/path/to/file) [program] [arguments]\n");
        exit(1);
        }
    if(!strncasecmp(argv[1],"--pid=",6)) {
        if(argv[2] == NULL) {
            printf(
                "Usage: duende (--pid=/path/to/file) [program] [arguments]\n");
            exit(1);
            }
        exec_argv_offset = 2;
        }

    /* Let children know that duende is running */
    if(setenv("DUENDE_IS_RUNNING","1",0) != 0) {
        printf("FATAL: Unable to set environment variable\n");
        exit(1);
        }

    /* The parent immediately exits */
    if(fork() != 0)
        exit(0);

    /* The child becomes a full-fledged daemon */
    setpgid(0,0); /* No longer visible in 'ps' without the 'auxw' argument */

    /* Write our PID to a file if the user so desires us to */
    if(exec_argv_offset == 2) {
        FILE *fp_pid = fopen(argv[1] + 6,"w");
        if(!fp_pid) {
            syslog(LOG_ALERT,"Fatal writing, to PID file, error\n");
            exit(1);
            }
        unsigned int local_pid = getpid();
        fprintf(fp_pid,"%u",local_pid);
        fclose(fp_pid);
        }

    /* Sysadmins expect HUP to reload, so we set that up */
    signal(SIGHUP,handle_hup);
    signal(SIGTERM,handle_term);
    signal(SIGINT,handle_term);

    pid = 0; log_pid = 0;

    for(;;) {
        if(pipe(stream1) != 0) {
            syslog(LOG_ALERT,"Fatal pipe error");
            exit(3);
            }
        pid = fork();
        if(pid == -1) {
            syslog(LOG_ALERT,"Fatal pid error");
            exit(1);
            }
        if(pid == 0) { /* Child; this one execs maradns */
            close(stream1[0]);
            /* Dup the standard output */
            if(dup2(stream1[1],1) != 1) {
                syslog(LOG_ALERT,"Fatal dup2 error 1");
                exit(4);
                }
            /* And the standard error */
            if(dup2(stream1[1],2) != 2) {
                syslog(LOG_ALERT,"Fatal dup2 error 2");
                exit(5);
                }
            argv[0] = argv[exec_argv_offset];
            execvp(argv[exec_argv_offset],argv + exec_argv_offset);
            /* OK, not found */
            printf("duende: %s: Command can't run, terminating\n",argv[exec_argv_offset]);
            syslog(LOG_ALERT,"Command can't run, terminating\n");
            exit(1);
            }

        /* Parent */
        close(stream1[1]);
        log_pid = fork();
        if(log_pid == 0) { /* Child to syslog all of MaraDNS' output */
            argv[0] = "duende-log-helper";
            log_helper(argv[exec_argv_offset],stream1[0]);
            syslog(LOG_ALERT,"log_helper finished, terminating\n");
            exit(1);
            }
        for(;;) {
            /* If we got a HUP signal, send it to the child */
            if(got_hup_signal == 1) {
                kill(pid,SIGHUP);
                got_hup_signal = 0;
                }
            /* If we got a TERM or INT signal, send it to the children
               then exit ourselves */
            else if(got_term_signal == 1) {
                /* XXX: make sure term really stops the children */
                kill(pid,SIGTERM);
                kill(log_pid,SIGTERM);
                syslog(LOG_ALERT,"got term signal, terminating\n");
                exit(0);
                }
            sleep(1);
            if(waitpid(pid,&exit_status,WNOHANG) == pid) { /* If child ended */
                handle_child_exited(exit_status,log_pid,pid);
                close(stream1[0]);
                break; /* Out of the inner loop; re-start Mara */
                }
            /* If logger terminated */
            if(waitpid(log_pid,&exit_status,WNOHANG) == log_pid) {
                handle_child_exited(exit_status,pid,log_pid);
                close(stream1[0]);
                break; /* Out of the inner loop; re-start Mara */
                }
            }
        }
    }

